### PR TITLE
Improve matches on command palette

### LINKED ISSUE #8184: command palette filter should trim spaces
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

1. Press cmd-shift-p
2. Press `space` 6 times
3. There are no matches

The main impact is obviously when spaces are typed in between search terms.

### Environment

Zed: v0.123.2 (Zed)
OS: macOS 14.3.1
Memory: 32 GiB
Architecture: aarch64

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your `~/Library/Logs/Zed/Zed.log` file to this issue.
If you only need the most recent lines, you can run the `zed: open log` command palette action to see the last 1000.

_No response_

### PR DESCRIPTION
**Pull request by Imgkl**
Release Notes:

- Fixed consecutive spaces in command palette influencing selection. #8184 

Optionally, include screenshots / media showcasing your addition that can be included in the release notes.

https://github.com/zed-industries/zed/assets/25414681/a4682247-f52c-4ab9-a32a-51ab5cf3dbcc

### CODE REVIEW

**Comment by mrnugget** [comment]
Hey! Thanks for opening the PR. I'm a bit concerned that this might change the command palette maybe too much, but maybe we can fix it in a smaller way.

Right now (on `main`) whitespace is significant in the palette, meaning: `<space><space><space>` will match entries that contain 3 spaces.

https://github.com/zed-industries/zed/assets/1185253/e29a33f5-6fd5-4da2-96a1-44be0d79ddc2

So maybe we want to keep the behaviour that `foo<space>bar<space>` matches entries with 2 spaces and `foo` and `bar` in it, by changing the code to only trim the start of the query?

**Comment by hferreiro** [comment]
> Right now (on main) whitespace is significant in the palette, meaning: <space><space><space> will match entries that contain 3 spaces.

Is that a common use case? I'm more concerned about accidentally typing 2 spaces between words and not finding some entries.

**Comment by mrnugget** [comment]
> Is that a common use case? I'm more concerned about accidentally typing 2 spaces between words and not finding some entries.

Not sure, to be honest, but it does allow you to type words partially and use the space.

Say I want to `toggle left dock`. Compare:

screenshot-2024-02-28-13 22 02@2x

vs

screenshot-2024-02-28-13 22 34@2x

**Comment by Imgkl** [comment]
Hey @mrnugget, 

Thanks for pointing why current implementation using whitespace is significant. I checked VS code, Looks like they do something similar to what you explained with "Toggle left" example. I'll tweak my code. 

@hferreiro, even I wasn't sure that this was a legit use case, but VS code do this.  Since more people are coming into Zed from VSC, it'll be good to keep this interaction similar.

**Comment by hferreiro** [comment]
I think I explained myself badly. I agree with you. My point was that if I type "tog  le" (2 spaces in between tog and le), those 2 spaces should be interpreted as 1; that is, any number of spaces is only meaningful in that they separate words.

**Comment by hferreiro** [comment]
In summary, I think we should trim spaces at the beginning and at the end of the search text, and also interpret any number of spaces between words as a single separator.

**Comment by Imgkl** [comment]
> we should trim spaces at the beginning and at the end of the search text, and also interpret any number of spaces between words as a single separator.


Now the changes are inline. 



image

File: crates/command_palette/src/command_palette.rs [hunk 1]
```diff
@@ -37,6 +39,10 @@ pub struct CommandPalette {
     picker: View<Picker<CommandPaletteDelegate>>,
 }
 
+lazy_static! {
+    static ref STRIP_UNNECESSARY_WHITESPACES: Regex = Regex::new(r"\s+").unwrap();
```
Comment (mrnugget) [inline crates/command_palette/src/command_palette.rs:43]: I'd prefer a name that describes what the regex matches, not what's being done when it's used:

```suggestion
    static ref CONSECUTIVE_WHITESPACES: Regex = Regex::new(r"\s+").unwrap();
```

**Review by mrnugget — APPROVED** [review]
Approving, but left one comment. Let me know if you want to address, otherwise I can hop on the branch too and do the rename.

File: crates/command_palette/src/command_palette.rs (hunk 1 above)
Comment (Imgkl) [reply crates/command_palette/src/command_palette.rs:43]: Oh it's fine @mrnugget. I'm doing it now. Let's close this one out properly. 😄

**Comment by TerminalFi** [comment]
Here is the code used in my testing


Regex method took: 979.481709ms
Manual method took: 162.163583ms
Manual method two took: 459.505292ms


```rust
use regex::Regex;
use std::time::Instant;

#[macro_use(lazy_static)]
extern crate lazy_static;

lazy_static! {
    static ref CONSECUTIVE_WHITESPACES: Regex = Regex::new(r"\s+").unwrap();
}
// Function using Regex
fn trim_consecutive_whitespaces_regex(input: &str) -> String {
    CONSECUTIVE_WHITESPACES
        .replace_all(input.trim(), " ")
        .to_string()
}

fn trim_consecutive_whitespaces_manual_two(input: &str) -> String {
    input
        .split_whitespace() //
        .collect::<Vec<&str>>()
        .join(" ")
}

fn trim_consecutive_whitespaces_manual(input: &str) -> String {
    let mut result = String::with_capacity(input.len());
    let mut last_char_was_whitespace = false;

    for char in input.trim().chars() {
        if char.is_whitespace() {
            if !last_char_was_whitespace {
                result.push(' ');
            }
            last_char_was_whitespace = true;
        } else {
            result.push(char);
            last_char_was_whitespace = false;
        }
    }
    result
}

fn main() {
    let test_str = "This   is a    test string    with  various  whitespaces   ";
    let iterations = 100_000;

    // Benchmark Regex method
    let start = Instant::now();
    for _ in 0..iterations {
        let _ = trim_consecutive_whitespaces_regex(test_str);
    }
    let duration_regex = start.elapsed();

    // Benchmark Manual method
    let start = Instant::now();
    for _ in 0..iterations {
        let _ = trim_consecutive_whitespaces_manual(test_str);
    }
    let duration_manual = start.elapsed();
    // Benchmark Manual method
    let start = Instant::now();
    for _ in 0..iterations {
        let _ = trim_consecutive_whitespaces_manual_two(test_str);
    }
    let duration_manual_two = start.elapsed();

    println!("Regex method took: {:?}", duration_regex);
    println!("Manual method took: {:?}", duration_manual);
    println!("Manual method two took: {:?}", duration_manual_two);
}
```

**Comment by mrnugget** [comment]
Thanks @TerminalFi! That makes sense. Felt a bit unsure about the regex myself. The only "problem" with your manual method I see is that `.is_whitespace` returns true for more than just ` `, so let's push the `char` instead of `' '` instead.

**Comment by TerminalFi** [comment]
@Imgkl Can you update the code to use ?

```rust
fn trim_consecutive_whitespaces_manual(input: &str) -> String {
    let mut result = String::with_capacity(input.len());
    let mut last_char_was_whitespace = false;

    for char in input.trim().chars() {
        if char.is_whitespace() {
            if !last_char_was_whitespace {
                result.push(char);
            }
            last_char_was_whitespace = true;
        } else {
            result.push(char);
            last_char_was_whitespace = false;
        }
    }
    result
}
```

**Comment by Imgkl** [comment]
@TerminalFi Yeah, pushing the changes in a bit.

**Review by mrnugget — APPROVED** [review]
Great! Let's get this one out the door :)

**Pull request merged**